storage/SMAIndexSubBlock.hpp

```diff
@@ -0,0 +1,418 @@
+ *          analyzed by the SMA.
+ */
+enum class Selectivity {
+  kAll,      // Select all tuples in storage subblock.
+  kSome,     // Select some of the tuples.
+  kNone,     // Select none of the tuples.
+  kUnknown,  // The predicate cannot be determined.
+  kUnsolved  // We have not tried to determine the selectivity.
+};
+
+/**
+ * @brief Helper method. Uses the stored values from the SMA Index to determine
+ *        selectivity of a predicate.
+ *
+ * @param literal The literal value to compare against.
+ * @param comparison The id of the predicate comparison function.
+ * @param min The minimum value associated with that attribute.
+ * @param max The maximum value associated with that attribute.
+ * @param less_comparator The less comparator for the attribute type.
+ * @param equals_comparator  The equals comparator for the attribute type.
+ * 
+ * @return Selectivity of this predicate.
+ */
+Selectivity getSelectivity(const TypedValue &literal,
+                           const ComparisonID comparison,
+                           const TypedValue &min,
+                           const TypedValue &max,
+                           const UncheckedComparator *less_comparator,
+                           const UncheckedComparator *equals_comparator);
+
+/**
+ * @brief A simple holding struct for a comparison predicate. Selectivity enum
+ *        indicates if the SMA has been used to solve the predicate and if so,
+ *        what is the selectivity over the block.
+ */
+struct SMAPredicate {
+  /**
+   * @brief Extracts a comparison predicate into an SMAPredicate.
+   *
+   * @param predicate A comparison of the form {attribute} {comparison} {literal}
+   *                  or {literal} {comparison} {attribute}.
+   * @return An SMAPredicate pointer which the caller must manage.
+   */
+  static SMAPredicate* ExtractSMAPredicate(const ComparisonPredicate& predicate);
+
+  const attribute_id attribute;
+  const ComparisonID comparison;
+  const TypedValue literal;
+  // How much of the TupleStore this predicate will select. kUnsolved indicates
+  // that the predicate had not been used previously.
+  Selectivity selectivity;
+
+ private:
+  SMAPredicate(const attribute_id attr,
+               const ComparisonID comp,
+               const TypedValue lit)
+      : attribute(attr),
+        comparison(comp),
+        literal(lit),
+        selectivity(Selectivity::kUnsolved) { }
+};
+
+//// Components of the index.
+
+// A 64-bit header.
+struct SMAHeader {
+  // Count refers the SQL aggregate COUNT.
+  std::uint32_t count;
+
+  union {
+    // Consistent refers to the index being in a consistent state.
+    bool consistent;
+    // Padding is so we keep this data structure to 64 bits in length.
+    std::uint32_t padding;
+  };
+};
+
+// Reference to an attribute value in a tuple. This struct is used to keep a
+// of the min and max value for an attribute.
+struct EntryReference {
+  tuple_id tuple;
```

Comment:
Can we have a more descriptive name here - something like entry_ref_tuple. Similarly for rest of the variables in this struct and SMAHeader.